File: 3/1.js
const fs = require('fs');

(function () {
  const contents = fs.readFileSync("input.txt").toString();
  let sum = contents.split("\r\n").map(line => {
    const first = line.slice(0, Math.ceil(line.length / 2)), last = line.slice(Math.floor(line.length / 2));

    let in_both = null;
    for (let a of first) {
      for (let b of last) {
        if (a === b) {
          in_both = a;
          break;
        }
      }
    }
    if (in_both) {
      let ascii = in_both.charCodeAt(0);
      let val = ascii - (ascii <= 90 ? 64 - 26 : 96);
      return val;
    }
    return -1;
  }).filter(x => x >= 0).reduce((a, b) => a + b, 0);
  console.log(sum);
})();

File: 3/hs/1.hs
import System.IO
import qualified Data.Text as Tx
import qualified Data.Text.Conversions as TxCv
import Data.List.Split
import Utils

main = do
  h <- openFile "../input.txt" ReadMode
  contents <- hGetContents h
  let
    dups =
      map (\[a,b] -> head $ head $ intersect a b) $ -- Get letters occuring in both lists. head of string returns char.
      map (\t -> map (\s -> tail $ splitOn "" $ Tx.unpack s) t) $ -- Split text into lists, remove first element ("")
      map (\ln -> let hlen = div (Tx.length ln) 2
        in [Tx.take hlen ln, Tx.drop hlen ln]) $ -- Split in half
      map (\ln -> TxCv.toText ln) $ -- Convert string to text
      lines contents -- Get lines (rounds)
    vals = map get_val dups
  print (sum vals)
  hClose h